commands2/sysid/sysidroutine.py

```diff
@@ -0,0 +1,175 @@
+        :param mechanism: Hardware interface for the SysId routine.
+        """
+        super().__init__(mechanism.subsystem.getName())
+        self.config = config
+        self.mechanism = mechanism
+        self.outputVolts = 0.0
+        self.logState = config.recordState or self.recordState
+
+    def quasistatic(self, direction: Direction) -> Command:
+        """Returns a command to run a quasistatic test in the specified direction.
+
+        The command will call the `drive` and `log` callbacks supplied at routine construction once
+        per iteration. Upon command end or interruption, the `drive` callback is called with a value of
+        0 volts.
+
+        :param direction: The direction in which to run the test.
+
+        :returns: A command to run the test.
+        """
+
+        timer = Timer()
+        if direction == self.Direction.kForward:
+            state = State.kQuasistaticForward
+        else:
+            state = State.kQuasistaticReverse
+
+        def execute():
+            self.outputVolts = direction.value * timer.get() * self.config.rampRate
+            self.mechanism.drive(self.outputVolts)
+            self.mechanism.log(self)
+            self.logState(state)
+
+        def end(interrupted: bool):
+            self.mechanism.drive(0.0)
+            self.logState(State.kNone)
+            timer.stop()
+
+        return (
+            self.mechanism.subsystem.runOnce(timer.start)
+            .andThen(self.mechanism.subsystem.run(execute))
+            .finallyDo(end)
+            .withName(
+                f"sysid-{SysIdRoutineLog.stateEnumToString(state)}-{self.mechanism.name}"
+            )
+            .withTimeout(self.config.timeout)
+        )
+
+    def dynamic(self, direction: Direction) -> Command:
+        """Returns a command to run a dynamic test in the specified direction.
+
+        The command will call the `drive` and `log` callbacks supplied at routine construction once
+        per iteration. Upon command end or interruption, the `drive` callback is called with a value of
+        0 volts.
+
+        :param direction: The direction in which to run the test.
+
+        :returns: A command to run the test.
+        """
+
+        if direction == self.Direction.kForward:
+            state = State.kDynamicForward
+        else:
+            state = State.kDynamicReverse
+
+        def command():
+            self.outputVolts = direction.value * self.config.stepVoltage
+
+        def execute():
+            self.mechanism.drive(self.outputVolts)
+            self.mechanism.log(self)
+            self.logState(state)
+
+        def end(interrupted: bool):
+            self.mechanism.drive(0.0)
+            self.logState(State.kNone)
+
+        return (
+            self.mechanism.subsystem.runOnce(command)
+            .andThen(self.mechanism.subsystem.run(execute))
+            .finallyDo(end)
+            .withName(f"sysid-{state}-{self.mechanism.name}")
```

Comment:
We'll want the same stringifier as above here.
```suggestion
            .withName(
                f"sysid-{SysIdRoutineLog.stateEnumToString(state)}-{self.mechanism.name}"
            )
```